experimental/web_client/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+#[async_trait(? Send)]
+impl UnaryClient for GrpcWebClient {
+    async fn message(&mut self, session_id: SessionId, body: Vec<u8>) -> anyhow::Result<Vec<u8>> {
+        let reply = grpc_web::grpc_web_unary::<UnaryRequest, UnaryResponse>(
+            &self.uri,
+            UnaryRequest {
+                session_id: session_id.to_vec(),
+                body,
+            },
+        )
+        .await?;
+        Ok(reply.body)
+    }
+}
+
+// The underlying [`GenericAttestationClient`] is held in a `wasm_mutex`, which
+// provides a queue of invocations. Ref: https://lib.rs/crates/wasm_mutex
+// TODO(#2908): Support concurrent invocations.
+type WebClientInner = Rc<wasm_mutex::Mutex<GenericAttestationClient<GrpcWebClient>>>;
+
+#[wasm_bindgen]
+pub struct WebClient {
+    inner: WebClientInner,
+}
+
+#[wasm_bindgen]
+impl WebClient {
+    pub async fn new(uri: String) -> Result<WebClient, String> {
+        let grpc_web_client = GrpcWebClient::create(&uri);
+        let inner = GenericAttestationClient::create(
+            grpc_web_client,
+            TEE_MEASUREMENT,
+            Box::new(|server_identity| {
+                if !server_identity.additional_info.is_empty() {
+                    Ok(())
+                } else {
+                    anyhow::bail!("No additional info provided.")
+                }
+            }),
+        )
+        .await
+        .context("Could not create Oak Functions client")
+        .map_err(|error| error.to_string())?;
+        let inner = Rc::new(wasm_mutex::Mutex::new(inner));
+        Ok(WebClient { inner })
+    }
+
+    // Typically this would an async function. Instead it's a sync function
+    // that returns a JS Promise (the JS equivalent of a rust future).
+    // The reason for this is that we cannot access the stack allocated self
+    // inside a Rust future.
+    // Ref: https://github.com/rustwasm/wasm-bindgen/issues/1858
+    pub fn invoke(&self, request: Vec<u8>) -> js_sys::Promise {
+        let inner = self.inner.clone();
+
+        wasm_bindgen_futures::future_to_promise(async move {
+            let response = WebClient::inner_invoke(inner, request)
+                .await
+                .map_err(|error| error.to_string())?;
+
+            // Get the response body, while removing trailing empty space.
+            let body = &response.body[0..response
+                .length
+                .try_into()
+                .map_err(|_| "Could not get response length")?];
+
+            // Construct a JavaScript object that contains the response status
+            // and body.
+            let js_response_object = js_sys::Object::new();
+            js_sys::Reflect::set(
+                &js_response_object,
+                &"status".into(),
+                &response.status.into(),
+            )?;
+            js_sys::Reflect::set(
+                &js_response_object,
+                &"body".into(),
+                &js_sys::Uint8Array::from(body),
```

Comment:
Curious, is there no way for wasm bindgen to do this, by generating some glue code from a Rust struct? Or using serde?

Reply:
wasm bindgen only binds methods of structs, not their values, see: https://rustwasm.github.io/wasm-bindgen/contributing/design/exporting-rust-struct.html

It does have bindings for serde that we could use: https://rustwasm.github.io/wasm-bindgen/reference/arbitrary-data-with-serde.html

Reply:
There's also a (third party) crate that does not rely on JSON serialization, but didn't look into it more, as this use-case is fairly simple.